### ISSUE TITLE
Proposal: Read KUBECONFIG environment variable as an alternative for --kubeconfig

### ISSUE THREAD

**Issue by kenfdev** [solo-io/glooctl#53]
Currently, it seems like the `KUBECONFIG` environment variable isn't being used.

I have a `KUBECONFIG` environment variable which targets a non-default config:

```
➜  echo $KUBECONFIG
/path/to/my/non-default/kube_config_cluster.yml
```

And `kubectl` targets that cluster:

```
➜ kubectl get svc -n gloo-system
NAME            TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                         AGE
control-plane   ClusterIP      10.233.15.12    <none>        8081/TCP                        22h
ingress         LoadBalancer   10.233.17.241   <pending>     8080:32198/TCP,8443:31533/TCP   22h
```

But `glooctl` showed `gloo-system-qloo-9090` which didn't make sense.

```
➜ glooctl upstream get
+--------------------------------+------------+--------+----------+
|              NAME              |    TYPE    | STATUS | FUNCTION |
+--------------------------------+------------+--------+----------+
| gloo-system-control-plane-8081 | kubernetes |        |          |
| gloo-system-ingress-8080       | kubernetes |        |          |
| gloo-system-ingress-8443       | kubernetes |        |          |
| gloo-system-qloo-9090          | kubernetes |        |          |
+--------------------------------+------------+--------+----------+
```

It was because `glooctl` was referencing a different cluster which was set in my `$HOME/.kube/config`.

I used the `--kubeconfig` option and it showed the upstreams nicely

```
➜ glooctl --kubeconfig /path/to/my/non-default/kube_config_cluster.yml upstream get
+---------------------------------------+------------+--------+----------+
|                 NAME                  |    TYPE    | STATUS | FUNCTION |
+---------------------------------------+------------+--------+----------+
| gloo-system-control-plane-8081        | kubernetes |        |          |
| gloo-system-ingress-8080              | kubernetes |        |          |
| gloo-system-ingress-8443              | kubernetes |        |          |
| ingress-nginx-default-http-backend-80 | kubernetes |        |          |
| openfaas-alertmanager-9093            | kubernetes |        |          |
| openfaas-alertmanager-external-9093   | kubernetes |        |          |
| openfaas-faas-netesd-8080             | kubernetes |        |          |
| openfaas-faas-netesd-external-8080    | kubernetes |        |          |
| openfaas-fn-url-ping-8080             | kubernetes |        |          |
| openfaas-gateway-8080                 | kubernetes |        | url-ping |
| openfaas-gateway-external-8080        | kubernetes |        |          |
| openfaas-nats-4222                    | kubernetes |        |          |
| openfaas-nats-external-4222           | kubernetes |        |          |
| openfaas-prometheus-9090              | kubernetes |        |          |
| openfaas-prometheus-external-9090     | kubernetes |        |          |
+---------------------------------------+------------+--------+----------+
```

It would be nice if `glooctl` respects the `KUBECONFIG` environment variable.

Thanks in advance.

**Comment by ilackarms** [comment]
i believe this can be done using `viper` (https://github.com/spf13/viper) which cobra (the cli library we use for `glooctl`) can leverage for easy conversion of flags to environment vars. if you'd like to look into a PR on this we'd be happy to merge it.

for now it should be suitable to provide the flag `--kubeconfig=${KUBECONFIG}`

**Comment by kenfdev** [comment]
Sure, I'll take a look at it and submit a PR.

**Comment by kenfdev** [comment]
@ilackarms I was expecting this to be pretty easy but have found out a bit more about the `KUBECONFIG` environment variable.

As explained in the [kubernetes docs](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#set-the-kubeconfig-environment-variable), `KUBECONFIG` can specify multiple config paths separated by `:` on mac and `;` on windows.

IIUC, instead of generating a restclient.Config [directly](https://github.com/solo-io/gloo/blob/master/pkg/bootstrap/configstorage/bootstrap.go#L26), maybe it is better to use the [ClientConfig](https://github.com/kubernetes/kubernetes/blob/625bce3ff6be2753cd9d27ef5fc2051cf2a45f85/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L56) of `k8s.io/client-go/tools/clientcmd`. This way [ClientConfigLoadingRules](https://github.com/kubernetes/kubernetes/blob/625bce3ff6be2753cd9d27ef5fc2051cf2a45f85/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L115) can be used and you can either set an [ExplicitPath](https://github.com/kubernetes/kubernetes/blob/625bce3ff6be2753cd9d27ef5fc2051cf2a45f85/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L116) or [Precedence](https://github.com/kubernetes/kubernetes/blob/625bce3ff6be2753cd9d27ef5fc2051cf2a45f85/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L117) for single or multiple config paths.

I've found that wercker/stern generates a client as explained above here:
https://github.com/wercker/stern/blob/9a597e725130db02421e7cb2046c33b0a3c11d9b/kubernetes/clientset.go#L31

And it is passed to kubernetes.NewForConfig like this:
https://github.com/wercker/stern/blob/9a597e725130db02421e7cb2046c33b0a3c11d9b/kubernetes/clientset.go#L55

This will affect `solo-io/gloo`'s code but in the long run, would be a better experience. WDYT?

Maybe start by supporting a single config file and later on, multiple config files would be better though.

**Comment by ilackarms** [comment]
yes! I am in favor of this change. Would you mind to do a PR to Gloo itself? (which glooctl shares for this piece of logic). this will actually help all of Gloo's components, not just `glooctl`.

thanks for this bit of research

**Comment by kenfdev** [comment]
Thanks again for the review! Sure! Let me have a look at the Gloo code as well and submit a PR 👍

**Comment by alexellis** [comment]
Can this issue be closed @kenfdev ?
